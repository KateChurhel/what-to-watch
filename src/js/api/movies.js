/* istanbul ignore file */
import { sendGetRequest } from './index';

export const getCategoryData = (category, data) => sendGetRequest(`/${category}/popular`, data);

export const getDetailsData = (category, id, data) => sendGetRequest(`/${category}/${id}`, data);

export const getSearchResult = (category, data) => sendGetRequest(`/search/${category}`, data);

export const getDiscoverResult = (data) => sendGetRequest('/discover/movie', data);

export const getGenres = () => sendGetRequest('/genre/movie/list');
