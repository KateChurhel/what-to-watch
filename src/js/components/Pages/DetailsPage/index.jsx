// libraries
import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { useParams } from 'react-router-dom';
// constants
import { MOVIE_TABS } from '../../../constants/categories';
import { URL_IMAGE } from '../../../constants/url';
// views
import Slider from '../../Partials/Slider';
import MovieTabOverview from '../../Partials/MovieTabOverview';
import MovieTabReviews from '../../Partials/MovieTabReviews';
import MovieTabDetails from '../../Partials/MovieTabDetails';
import MovieTabCast from '../../Partials/MovieTabCast';

const DetailsPage = (props) => {
  const {
    formattedRunTime, releaseYear, genresList, cardTitle, reviews = [], tagline,
    poster, backgroundImage, voteAverage, voteCount, overview, cast = [], recommendations = [],
  } = props;
  const [activeTab, setActiveTab] = useState(MOVIE_TABS.OVERVIEW);
  const { category } = useParams();

  const renderActiveTab = () => {
    switch (activeTab) {
      case MOVIE_TABS.DETAILS:
        return (
          <MovieTabDetails {...props} />
        );
      case MOVIE_TABS.REVIEWS:
        return (
          <MovieTabReviews reviews={reviews} />
        );
      case MOVIE_TABS.CAST:
        return (
          <MovieTabCast cast={cast} />
        );
      default:
        return (
          <MovieTabOverview
            cast={cast}
            overview={overview}
            voteAverage={voteAverage}
            voteCount={voteCount}
          />
        );
    }
  };

  return (
    <>
      <section className="movie-card movie-card--full">
        <div className="movie-card__hero">
          <div
            className="movie-card__bg"
            style={{ backgroundImage: `url('${URL_IMAGE.backgroundImage}${backgroundImage}')` }}
          />
          <div className="container">
            <div className=" movie-card__wrap">
              <div className="movie-card__desc">
                <h2 className="movie-card__title">{`${cardTitle} (${releaseYear})`}</h2>
                <p className="movie-card__meta">
                  <span className="movie-card__genre">{genresList}</span>
                  <span className="movie-card__runtime">{formattedRunTime}</span>
                </p>
                <i className="movie-card__tagline">{tagline}</i>
              </div>
            </div>
          </div>
        </div>
        <div className="container">
          <div className=" movie-card__wrap movie-card__translate-top">
            <div className="movie-card__info">
              <div className="movie-card__poster movie-card__poster--big">
                <img
                  alt={cardTitle}
                  height="327"
                  loading="lazy"
                  src={`${URL_IMAGE.poster}${poster}`}
                  width="218"
                />
              </div>

              <div className="movie-card__desc">
                <nav className="movie-nav movie-card__nav">
                  <ul className="movie-nav__list">
                    {Object.values(MOVIE_TABS)
                      .map((tab) => (
                        <li
                          key={tab}
                          className={`movie-nav__item ${tab === activeTab ? 'movie-nav__item--active' : ''}`}
                        >
                          <button
                            className="movie-nav__link"
                            onClick={(event) => {
                              event.preventDefault();
                              setActiveTab(tab);
                            }}
                            type="button"
                          >
                            {tab}
                          </button>
                        </li>
                      ))}
                  </ul>
                </nav>

                <div className="movie-card__tab-wrapper">
                  {renderActiveTab()}
                </div>
              </div>
            </div>
          </div>
        </div>
      </section>
      {recommendations.length ? (
        <div className="container">
          <section className="catalog catalog--like-this">
            <h2 className="catalog__title">More like this</h2>
            <Slider category={category} items={recommendations} />
          </section>
        </div>
      ) : null}
    </>
  );
};

DetailsPage.propTypes = {
  formattedRunTime: PropTypes.string,
  releaseYear: PropTypes.string,
  genresList: PropTypes.string,
  cardTitle: PropTypes.string,
  reviews: PropTypes.arrayOf(PropTypes.object),
  tagline: PropTypes.string,
  poster: PropTypes.string,
  backgroundImage: PropTypes.string,
  voteAverage: PropTypes.number,
  voteCount: PropTypes.number,
  overview: PropTypes.string,
  cast: PropTypes.arrayOf(PropTypes.object),
  recommendations: PropTypes.arrayOf(PropTypes.object),
};

export default DetailsPage;
