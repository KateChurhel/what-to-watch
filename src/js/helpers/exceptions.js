// constants
import DEFAULT_ERROR from '../constants/errors';

export class RequestError extends Error {
  constructor(status = '', responseData) {
    super();

    const errorData = responseData || {};

    this.message = errorData.message || DEFAULT_ERROR;
    this.errors = errorData.errors;
    this.status = status;
    this.name = 'RequestError';
  }
}

export default {};
