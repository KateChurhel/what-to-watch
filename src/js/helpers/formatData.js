import moment from 'moment';
import 'moment-duration-format';

export const formatTime = (time, unit, format) => moment.duration(time, unit).format(format, { trim: false });
export const formatDate = (date, format) => moment(date).format(format);

export const parseQueryString = (queryString) => {
  const query = {};
  const pairs = (queryString[0] === '?' ? queryString.substr(1) : queryString).split('&');
  for (let i = 0; i < pairs.length; i++) {
    const pair = pairs[i].split('=');
    query[decodeURIComponent(pair[0])] = decodeURIComponent(pair[1] || '');
  }

  return query;
};

export const getStringValues = (array = [], key) => (array.length ? array.map((item) => item[key]).join(', ') : '');

export const formatCardData = ({
  number_of_episodes: numberOfEpisodes, number_of_seasons: numberOfSeasons, original_name: originalName,
  episode_run_time: episodeRunTime, reviews: { results: reviews = [] } = {}, homepage, original_title: originalTitle,
  original_language: originalLanguage, spoken_languages: spokenLanguages, status, runtime, tagline,
  release_date: releaseDate, genres, poster_path: poster, backdrop_path: backgroundImage,
  vote_average: voteAverage, vote_count: voteCount, overview, title, name, id,
  credits: { cast = [] } = {}, recommendations: { results: recommendations = [] } = {},
} = {}) => {
  const formattedRunTime = formatTime(runtime || episodeRunTime, 'minutes', 'h[h] m[m]');
  const releaseYear = formatDate(releaseDate, 'YYYY');
  const genresList = getStringValues(genres, 'name');
  const spokenLanguagesList = getStringValues(spokenLanguages, 'name');

  return {
    id,
    numberOfEpisodes,
    numberOfSeasons,
    episodeRunTime,
    reviews,
    homepage,
    originalLanguage,
    status,
    tagline,
    releaseDate,
    poster,
    backgroundImage,
    voteAverage,
    voteCount,
    overview,
    cast,
    recommendations,
    originalTitle: originalTitle || originalName,
    cardTitle: title || name,
    formattedRunTime,
    releaseYear,
    genresList,
    spokenLanguagesList,
  };
};
