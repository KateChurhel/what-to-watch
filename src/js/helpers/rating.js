import USER_RATINGS from '../constants/rating';

export const computeRatingLevel = (rating) => USER_RATINGS.filter(
  (rank) => rank.min <= rating,
).pop().title;
