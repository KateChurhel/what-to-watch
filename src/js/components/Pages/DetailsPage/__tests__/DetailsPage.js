import React from 'react';
import { act, render } from '@testing-library/react';
import { MemoryRouter, Route } from 'react-router-dom';
import { mount } from 'enzyme';
import DetailsPage from '../index';
import CATEGORIES, { MOVIE_TABS } from '../../../../constants/categories';
import ROUTES from '../../../../constants/routes';
import { MOCK_MOVIES } from '../../../../mocks/movies';
import { formatCardData } from '../../../../helpers/formatData';

describe('Details Page: ', () => {
  it('correctly renders', async () => {
    const { asFragment } = render(
      <MemoryRouter initialEntries={[`/details/${CATEGORIES.movie}/1`]} initialIndex={1}>
        <Route path={ROUTES.detailsPage}>
          <DetailsPage {...formatCardData(MOCK_MOVIES[0])} />
        </Route>
      </MemoryRouter>,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('correctly renders with empty result', async () => {
    const { asFragment } = render(
      <MemoryRouter initialEntries={[`/details/${CATEGORIES.movie}/1`]} initialIndex={1}>
        <Route path={ROUTES.detailsPage}>
          <DetailsPage />
        </Route>
      </MemoryRouter>,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  describe('', () => {
    let wrapper;

    beforeEach(async () => {
      await act(async () => {
        wrapper = await mount(<MemoryRouter initialEntries={[`/details/${CATEGORIES.movie}/1`]} initialIndex={1}>
          <Route path={ROUTES.detailsPage}>
            <DetailsPage {...formatCardData(MOCK_MOVIES[0])} />
          </Route>
                              </MemoryRouter>);
      });
      wrapper.update();
    });

    it('open details tab', async () => {
      const link = wrapper.findWhere((node) => (
        node.name() === 'button'
          && node.text() === MOVIE_TABS.DETAILS
      ));
      link.simulate('click');
      expect(wrapper.find('.movie-nav__item--active button').props().children).toEqual(MOVIE_TABS.DETAILS);
    });

    it('open reviews tab', async () => {
      const link = wrapper.findWhere((node) => (
        node.name() === 'button'
          && node.text() === MOVIE_TABS.REVIEWS
      ));
      link.simulate('click');
      expect(wrapper.find('.movie-nav__item--active button').props().children).toEqual(MOVIE_TABS.REVIEWS);
    });

    it('open cast tab', async () => {
      const link = wrapper.findWhere((node) => (
        node.name() === 'button'
          && node.text() === MOVIE_TABS.CAST
      ));
      link.simulate('click');
      expect(wrapper.find('.movie-nav__item--active button').props().children).toEqual(MOVIE_TABS.CAST);
    });
  });
});
