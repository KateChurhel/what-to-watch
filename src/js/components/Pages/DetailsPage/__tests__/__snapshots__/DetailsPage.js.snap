// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Details Page:  correctly renders 1`] = `
<DocumentFragment>
  <section
    class="movie-card movie-card--full"
  >
    <div
      class="movie-card__hero"
    >
      <div
        class="movie-card__bg"
        style="background-image: url(https://image.tmdb.org/t/p/original/qJeU7KM4nT2C1WpOrwPcSDGFUWE.jpg);"
      />
      <div
        class="container"
      >
        <div
          class=" movie-card__wrap"
        >
          <div
            class="movie-card__desc"
          >
            <h2
              class="movie-card__title"
            >
              La La Land (2016)
            </h2>
            <p
              class="movie-card__meta"
            >
              <span
                class="movie-card__genre"
              >
                Comedy, Drama, Romance, Music
              </span>
              <span
                class="movie-card__runtime"
              >
                2h 9m
              </span>
            </p>
            <i
              class="movie-card__tagline"
            >
              Here's to the fools who dream.
            </i>
          </div>
        </div>
      </div>
    </div>
    <div
      class="container"
    >
      <div
        class=" movie-card__wrap movie-card__translate-top"
      >
        <div
          class="movie-card__info"
        >
          <div
            class="movie-card__poster movie-card__poster--big"
          >
            <img
              alt="La La Land"
              height="327"
              loading="lazy"
              src="https://image.tmdb.org/t/p/w300/uDO8zWDhfWwoFdKS4fzkUJt0Rf0.jpg"
              width="218"
            />
          </div>
          <div
            class="movie-card__desc"
          >
            <nav
              class="movie-nav movie-card__nav"
            >
              <ul
                class="movie-nav__list"
              >
                <li
                  class="movie-nav__item movie-nav__item--active"
                >
                  <button
                    class="movie-nav__link"
                    type="button"
                  >
                    Overview
                  </button>
                </li>
                <li
                  class="movie-nav__item "
                >
                  <button
                    class="movie-nav__link"
                    type="button"
                  >
                    Details
                  </button>
                </li>
                <li
                  class="movie-nav__item "
                >
                  <button
                    class="movie-nav__link"
                    type="button"
                  >
                    Reviews
                  </button>
                </li>
                <li
                  class="movie-nav__item "
                >
                  <button
                    class="movie-nav__link"
                    type="button"
                  >
                    Cast
                  </button>
                </li>
              </ul>
            </nav>
            <div
              class="movie-card__tab-wrapper"
            >
              <div
                class="movie-rating"
              >
                <div
                  class="movie-rating__score"
                >
                  7.9
                </div>
                <p
                  class="movie-rating__meta"
                >
                  <span
                    class="movie-rating__level"
                  >
                    Good
                  </span>
                  <span
                    class="movie-rating__count"
                  >
                    12161 ratings
                  </span>
                </p>
              </div>
              <p
                class="movie-card__text"
              >
                Mia, an aspiring actress, serves lattes to movie stars in between auditions and Sebastian, a jazz musician, scrapes by playing cocktail party gigs in dingy bars, but as success mounts they are faced with decisions that begin to fray the fragile fabric of their love affair, and the dreams they worked so hard to maintain in each other threaten to rip them apart.
              </p>
            </div>
          </div>
        </div>
      </div>
    </div>
  </section>
  <div
    class="container"
  >
    <section
      class="catalog catalog--like-this"
    >
      <h2
        class="catalog__title"
      >
        More like this
      </h2>
      <div
        class="swiper-container swiper-container-initialized swiper-container-horizontal"
      >
        <div
          class="swiper-button-prev"
        />
        <div
          class="swiper-button-next"
        />
        <div
          class="swiper-wrapper"
          style="transition: 0ms; transform: translate3d(NaNpx, 0px, 0px);"
        >
          <div
            class="swiper-slide"
          >
            <a
              class="small-movie-card catalog__movies-card small-movie-card__link"
              href="/details/movie/340666"
            >
              <div
                class="small-movie-card__image "
              >
                <img
                  alt="Nocturnal Animals"
                  height="450"
                  loading="lazy"
                  src="https://image.tmdb.org/t/p/w300/mdLDgQBD0va09npSQX5Zgo2evXM.jpg"
                  width="300"
                />
              </div>
              <div
                class="small-movie-card__info"
              >
                <h3
                  class="small-movie-card__title"
                >
                  Nocturnal Animals
                </h3>
              </div>
            </a>
          </div>
          <div
            class="swiper-slide"
          >
            <a
              class="small-movie-card catalog__movies-card small-movie-card__link"
              href="/details/movie/314365"
            >
              <div
                class="small-movie-card__image "
              >
                <img
                  alt="Spotlight"
                  height="450"
                  loading="lazy"
                  src="https://image.tmdb.org/t/p/w300/pt7ZeF0RKJZgFe11YUxaO4J8GCc.jpg"
                  width="300"
                />
              </div>
              <div
                class="small-movie-card__info"
              >
                <h3
                  class="small-movie-card__title"
                >
                  Spotlight
                </h3>
              </div>
            </a>
          </div>
          <div
            class="swiper-slide"
          >
            <a
              class="small-movie-card catalog__movies-card small-movie-card__link"
              href="/details/movie/381284"
            >
              <div
                class="small-movie-card__image "
              >
                <img
                  alt="Hidden Figures"
                  height="450"
                  loading="lazy"
                  src="https://image.tmdb.org/t/p/w300/9lfz2W2uGjyow3am00rsPJ8iOyq.jpg"
                  width="300"
                />
              </div>
              <div
                class="small-movie-card__info"
              >
                <h3
                  class="small-movie-card__title"
                >
                  Hidden Figures
                </h3>
              </div>
            </a>
          </div>
        </div>
      </div>
    </section>
  </div>
</DocumentFragment>
`;

exports[`Details Page:  correctly renders with empty result 1`] = `
<DocumentFragment>
  <section
    class="movie-card movie-card--full"
  >
    <div
      class="movie-card__hero"
    >
      <div
        class="movie-card__bg"
        style="background-image: url(https://image.tmdb.org/t/p/originalundefined);"
      />
      <div
        class="container"
      >
        <div
          class=" movie-card__wrap"
        >
          <div
            class="movie-card__desc"
          >
            <h2
              class="movie-card__title"
            >
              undefined (undefined)
            </h2>
            <p
              class="movie-card__meta"
            >
              <span
                class="movie-card__genre"
              />
              <span
                class="movie-card__runtime"
              />
            </p>
            <i
              class="movie-card__tagline"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      class="container"
    >
      <div
        class=" movie-card__wrap movie-card__translate-top"
      >
        <div
          class="movie-card__info"
        >
          <div
            class="movie-card__poster movie-card__poster--big"
          >
            <img
              height="327"
              loading="lazy"
              src="https://image.tmdb.org/t/p/w300undefined"
              width="218"
            />
          </div>
          <div
            class="movie-card__desc"
          >
            <nav
              class="movie-nav movie-card__nav"
            >
              <ul
                class="movie-nav__list"
              >
                <li
                  class="movie-nav__item movie-nav__item--active"
                >
                  <button
                    class="movie-nav__link"
                    type="button"
                  >
                    Overview
                  </button>
                </li>
                <li
                  class="movie-nav__item "
                >
                  <button
                    class="movie-nav__link"
                    type="button"
                  >
                    Details
                  </button>
                </li>
                <li
                  class="movie-nav__item "
                >
                  <button
                    class="movie-nav__link"
                    type="button"
                  >
                    Reviews
                  </button>
                </li>
                <li
                  class="movie-nav__item "
                >
                  <button
                    class="movie-nav__link"
                    type="button"
                  >
                    Cast
                  </button>
                </li>
              </ul>
            </nav>
            <div
              class="movie-card__tab-wrapper"
            >
              <div
                class="movie-rating"
              >
                <div
                  class="movie-rating__score"
                >
                  0
                </div>
                <p
                  class="movie-rating__meta"
                >
                  <span
                    class="movie-rating__level"
                  >
                    Bad
                  </span>
                  <span
                    class="movie-rating__count"
                  >
                    0 ratings
                  </span>
                </p>
              </div>
              <p
                class="movie-card__text"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </section>
</DocumentFragment>
`;
